nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py

```diff
@@ -272,21 +278,21 @@ def get_loss_and_metrics(self, batch, forward_only):
             rewards_all_mean = rewards_all.mean()
             rewards_all_std = rewards_all.std()
 
-            # average loss across micro batches
-            loss_tensors_list = [loss_reduced["avg"] for loss_reduced in losses_reduced_per_micro_batch]
-            loss_tensor = torch.concat(loss_tensors_list)
-            loss_mean = loss_tensor.mean()
-            acc_tensors_list = [loss_reduced["acc"] for loss_reduced in losses_reduced_per_micro_batch]
-
-            if len(acc_tensors_list) == 1:
-                acc_tensor = acc_tensors_list[0]
-            elif len(acc_tensors_list) > 1:
-                acc_tensor = torch.concat(acc_tensors_list)
-            acc_mean = acc_tensor.mean()
+            num_valid_pairs = torch.stack(
+                [loss_reduced["num_valid_pairs"] for loss_reduced in losses_reduced_per_micro_batch]
+            )
+            loss_tensor = torch.cat([loss_reduced["avg"].view(1) for loss_reduced in losses_reduced_per_micro_batch])
+            acc_tensor = torch.cat([loss_reduced["acc"].view(1) for loss_reduced in losses_reduced_per_micro_batch])
+
+            loss_mean = (loss_tensor * num_valid_pairs).sum() / num_valid_pairs.sum()
+            acc_mean = (acc_tensor * num_valid_pairs).sum() / num_valid_pairs.sum()
+            weights = num_valid_pairs.sum()
```

Comment:
Minor rewrite but also casting `weights` to float as I suspect this may otherwise cause a hang with PP>1 when we broadcast the weights below:
```suggestion
            weights = num_valid_pairs.sum().float()
            loss_mean = (loss_tensor * num_valid_pairs).sum() / weights
            acc_mean = (acc_tensor * num_valid_pairs).sum() / weights
```

Reply:
Good point, thanks for flagging